task/move.go

```diff
@@ -86,27 +87,79 @@ func Move(fetch Fetcher, put Putter) {
 	}
 }
 
+func (o *oldStore) fetchB() (thing Thing, ok bool, err error) {
+	// simulate the delay
+	time.Sleep(2 * time.Second)
+	fmt.Println("fetchB: fetching...", o.bookNo)
+
+	if o.bookNo == 5 {
+		return thing, ok, errors.New("Fetch encountered network issues")
+	}
+
+	if o.bookNo > len(o.bookInventory)-1 {
+
+		return thing, ok, err
+	}
+
+	time.Sleep(1 * time.Second)
+	thing, ok = o.bookInventory[o.bookNo], true
+	o.bookNo++
+	return
+}
+
+func (n *newStore) putB(thing Thing) error {
+	time.Sleep(5 * time.Second)
+
+	n.inventory = append(n.inventory, thing)
+
+	if len(n.inventory) == 4 {
+		return errors.New("store is filled up")
+	}
+
+	return nil
+}
+
 // MaybeMove is exactly the same as Move except that it may return an error
 // because fetch() and put() may return errors. If no errors occur then
 // MaybeMove returns under the same conditions as Move(). If an error occurs
 // then MaybeMove returns earlier even if there are more Things to fetch().
 func MaybeMove(fetch MaybeFetcher, put MaybePutter) error {
-	// [TODO]
+	ch := make(chan Thing)
+	errCh := make(chan error, 1)
+	done := make(chan struct{}, 1)
+	defer close(done)
 
 	go func() {
-		// [TODO]
-		t, ok, err := fetch()
-		_, _, _ = t, ok, err // [Remove later]
-		// [TODO]
-	}()
+		defer close(errCh)
```

Comment:
There is a subtle bug here that I will hint to with questions rather than explaining. What happens if both `fetch()` and `put()` have an error at approximately the same time?